network_snapshot/network_snapshot.py

```diff
@@ -0,0 +1,441 @@
+    return ((epoch + 1) * blocks_per_epoch) - 1
+
+def get_epoch_first_block(epoch, blocks_per_epoch):
+    return epoch * blocks_per_epoch
+
+def init_loggers():
+    # Log to dump all blocks in chain
+    logger = logging.getLogger('blockchain')
+    logger.addHandler(get_simple_rotating_log_handler(block_log))
+
+    # Log to dump all RPC results from account snapshots
+    logger = logging.getLogger('rpc')
+    logger.addHandler(get_simple_rotating_log_handler(rpc_log))
+
+    # Log to dump all RPC results from validator snapshots
+    logger = logging.getLogger('validator')
+    logger.addHandler(get_simple_rotating_log_handler(validator_log))
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('endpoint', help = 'Endpoint to query for blockchain data')
+    parser.add_argument('--retries', default = 5, help = 'Number of retries for a request before failing')
+    parser.add_argument('--delay', default = 30, type = int, help = 'Delay between iterations in seconds')
+    parser.add_argument('--verbose', default = False, action = 'store_true', help = 'Verbose print for debug')
+
+    args = parser.parse_args()
+
+    if args.verbose:
+        def v_print(s):
+            print(s)
+    else:
+        def v_print(s):
+            return
+
+    network_accounts = set()
+    prev_epoch = 0
+    prev_block = 0
+
+    blocks_per_epoch = 0
+    metadata = get_node_metadata(args.endpoint, args.retries)
+    try:
+        blocks_per_epoch = metadata['blocks-per-epoch']
+    except Exception as e:
+        print("Error getting blocks_per_epoch. Check network status?")
+        exit()
+    v_print(f'Blocks Per Epoch: {blocks_per_epoch}')
+
+    # NOTE: Should not fail
+    if not path.exists(data):
+        os.mkdir(data)
+    else:
+        # Read last existing accounts file for epoch & active network accounts
+        account_files = glob(path.join(data, 'accounts*.csv'))
+        if len(account_files) > 0:
+            # Might need to use min to get latest file
+            latest_account_file = max(account_files, key = path.getctime)
+            v_print(f'Reading initial data from: {latest_account_file}')
+            with open(latest_account_file, 'r', encoding = 'utf8') as f:
+                reader = csv.reader(f)
+                # Skip header line
+                next(reader)
+                for line in reader:
+                    network_accounts.add(line[0].strip())
+            try:
+                epoch_token = path.basename(latest_account_file).split('.')[0].split('_')[1]
+                prev_epoch = int(epoch_token)
+                # Since account filed generated, all blocks in that epoch should have been fetched
+                prev_block = get_epoch_first_block(int(epoch_token) + 1, blocks_per_epoch)
+            except Exception:
+                pass
+
+    # Create loggers after creating data directory
+    init_loggers()
+
+    while True:
```

Comment:
don't in the script, use a cronjob or a jenkins job to have a better control and test.